### PR TITLE
Extend block dump

### PR DESCRIPTION
**Pull request by Sventimir**
Extend the `dump_blocks` with:
* proper CLI
* generate full blocks rather than just precomputed values
* convert between formats: JSON, sexp and binary (although with some caveats).
* output directly to files
* inject parent state hash into generated blocks (although this is done the dumb way of simply replacing that field in the block's data structure).

Explain how you tested your changes:
* run the binary with various option configurations; observe the results
* pass the generated blocks through `delegation_verify` app
* make sure that with no CLI options the app behaves the same as previously.

Caveats:
* some encoding combinations don't work properly (no JSON encoding for full blocks; inability to load sexp dump of a full block due to opacity of some values)
* `delegation_verfiy` only accepts the block if `--no-check` is enabled, probably because some protocol constants are being randomised.

Checklist:

- [x] Dependency versions are unchanged
  - Notify Velocity team if dependencies must change in CI
- [x] Modified the current draft of release notes with details on what is completed or incomplete within this project
- [x] Document code purpose, how to use it
  - Mention expected invariants, implicit constraints
- [x] Tests were added for the new behavior
  - Document test purpose, significance of failures
  - Test names should reflect their purpose
- [x] All tests pass (CI will check this if you didn't)
- [x] Serialized types are in stable-versioned modules

### CODE REVIEW

**Comment by Sventimir** [comment]
!ci-build-me

**Comment by Sventimir** [comment]
!ci-build-me

**Comment by Sventimir** [comment]
!ci-build-me

File: src/app/dump_blocks/dump_blocks.ml [hunk 1]
```diff
@@ -12,43 +67,104 @@ let precomputed_values = Lazy.force Precomputed_values.for_unit_tests
...
+  ]
+
+let command =
+  Command.basic
+    ~summary:"Transcribe between block formats or generate a random block"
+    (let%map_open.Command outputs =
+       flag
+         "-o"
+         (listed encoded_block)
+         ~doc:"output <format>:<path>; formats are: sexp / json / bin; a dash (-) dentoes stdout"
```
Comment (emberian) [inline src/app/dump_blocks/dump_blocks.ml:121]:
```suggestion
         ~doc:"output <format>:<path>; formats are: sexp / json / bin; a dash (-) denotes stdout (example: -o json:-)"
```

File: src/app/dump_blocks/dump_blocks.ml [hunk 2]
```diff
@@ -12,43 +67,104 @@ let precomputed_values = Lazy.force Precomputed_values.for_unit_tests
...
+     and input =
+       flag
+         "-i"
+         (optional encoded_block)
+         ~doc:"input <format>:<path>; formats are: sexp / json / bin; a dash (-) dentoes stdin"
```
Comment (emberian) [inline src/app/dump_blocks/dump_blocks.ml:126]:
```suggestion
         ~doc:"input <format>:<path>; formats are: sexp / json / bin; a dash (-) denotes stdin (example: -i json:-)"
```

**Review by emberian — APPROVED** [review]
Looks great modulo denotes typo!

**Comment by Sventimir** [comment]
!ci-build-me

**Comment by Sventimir** [comment]
!ci-build-me

**Comment by Sventimir** [comment]
!ci-build-me

File: src/app/dump_blocks/dump_blocks.ml [hunk 3]
```diff
@@ -1,6 +1,61 @@
...
+  | (Block, { encoding = Sexp; filename }) ->
+     { encoding = (module Encoding.Sexp_block)
+     ; filename
+     }
+  | (Block, { encoding = Json; filename }) ->
+     failwith "Json encoding for full blocks not supported."
+  | (Block, { encoding = Binary; filename }) ->
+     { encoding = (module Encoding.Binary_block)
+     ; filename
+     }
```
Comment (joaosreis) [inline src/app/dump_blocks/dump_blocks.ml:43]:
```suggestion
let mk_io : type a. a Encoding.content -> Encoding.t io -> a codec io =
 fun content io ->
  let open Encoding in
  { io with
    encoding =
      ( match (content, io.encoding) with
      | Precomputed, Sexp ->
          (module Sexp_precomputed)
      | Precomputed, Json ->
          (module Json_precomputed)
      | Precomputed, Binary ->
          (module Binary_precomputed)
      | Block, Sexp ->
          (module Sexp_block)
      | Block, Json ->
          failwith "Json encoding for full blocks not supported."
      | Block, Binary ->
          (module Binary_block) )
  }
```

File: src/app/dump_blocks/dump_blocks.ml [hunk 4]
```diff
@@ -12,43 +67,104 @@ let precomputed_values = Lazy.force Precomputed_values.for_unit_tests
...
+       | [] -> default_outputs
+       | _ -> outputs
+     in
+     let conf =
+       if full then
+         Conf ( Option.map input ~f:(fun i -> mk_io (Block, i))
+              , List.map ~f:mk_io (List.map ~f:(fun x -> (Encoding.Block, x)) outs))
+       else
+         Conf ( Option.map input ~f:(fun i -> mk_io (Precomputed, i))
+              , List.map ~f:mk_io (List.map ~f:(fun x -> (Encoding.Precomputed, x)) outs))
```
Comment (joaosreis) [inline src/app/dump_blocks/dump_blocks.ml:149]:
```suggestion
         Conf ( Option.map input ~f:(mk_io Block)
              , List.map ~f:(mk_io Encoding.Block) outs)
       else
         Conf ( Option.map input ~f:(mk_io Precomputed)
              , List.map ~f:(mk_io Encoding.Precomputed) outs)
```

**Review by joaosreis — APPROVED** [review]
LGTM, approved. Just left an optional suggestion, if you feel like it simplifies that piece of code.

**Comment by Sventimir** [comment]
Good point, @joaosreis, thanks!

**Comment by Sventimir** [comment]
!ci-build-me

**Comment by emberian** [comment]
!ci-build-me

**Comment by Sventimir** [comment]
!ci-build-me

**Comment by Sventimir** [comment]
!ci-build-me

**Comment by Sventimir** [comment]
!ci-build-me

**Review by dkijania — APPROVED** [review]

**Comment by deepthiskumar** [comment]
@Sventimir where does this tool get used?

**Comment by Sventimir** [comment]
!ci-build-me

**Comment by Sventimir** [comment]
@deepthiskumar, we're using it to generate dummy blocks to test our delegation program tooling. The idea is that we don't want to use real nodes in the process, because the generate unpredictability due to p2p networking. Still, we need a blockchain to submit to the uptime service, so need arises to generate dummy blocks just so that they reference each other to form a chain. This tool does the job.

**Pull request merged**